File: app/src/main/java/io/github/sithengineer/motoqueiro/app/AppModule.java
package io.github.sithengineer.motoqueiro.app;

import android.app.Application;
import android.preference.PreferenceManager;
import dagger.Module;
import dagger.Provides;
import io.github.sithengineer.motoqueiro.authentication.AccountManager;
import io.github.sithengineer.motoqueiro.util.VariableScrambler;
import javax.inject.Singleton;

@Module public class AppModule {
  private final Application application;
  private final android.accounts.AccountManager androidAccountManager;
  private final String mySecretKey;

  public AppModule(Application application,
      android.accounts.AccountManager accountManager, String mySecretKey) {
    this.application = application;
    this.androidAccountManager = accountManager;
    this.mySecretKey = mySecretKey;
  }

  @Provides @Singleton Application providesApplication() {
    return application;
  }

  @Provides @Singleton Preferences providesSharedPrefs() {
    return new Preferences(PreferenceManager.getDefaultSharedPreferences(application));
  }

  @Provides @Singleton VariableScrambler providesVariableScrambler() {
    return new VariableScrambler(mySecretKey);
  }

  @Provides @Singleton AccountManager providesAccountManager() {
    return new AccountManager(androidAccountManager);
  }
}

File: app/src/main/java/io/github/sithengineer/motoqueiro/app/AppModule.kt
package io.github.sithengineer.motoqueiro.app

import android.app.Application
import android.arch.persistence.room.Room
import android.preference.PreferenceManager
import dagger.Module
import dagger.Provides
import io.github.sithengineer.motoqueiro.authentication.AccountManager
import io.github.sithengineer.motoqueiro.data.local.AppDatabase
import io.github.sithengineer.motoqueiro.scope.ApplicationScope
import io.github.sithengineer.motoqueiro.util.VariableScrambler
import io.reactivex.Scheduler
import io.reactivex.schedulers.Schedulers

@Module
class AppModule(private val application: Application,
    private val androidAccountManager: android.accounts.AccountManager,
    private val mySecretKey: String) {

  @Provides
  @ApplicationScope
  internal fun providesApplication(): Application {
    return application
  }

  @Provides
  @ApplicationScope
  internal fun providesIoScheduler(): Scheduler {
    return Schedulers.io()
  }

  @Provides
  @ApplicationScope
  internal fun providesSharedPrefs(): Preferences {
    return Preferences(PreferenceManager.getDefaultSharedPreferences(application))
  }

  @Provides
  @ApplicationScope
  internal fun providesVariableScrambler(): VariableScrambler {
    return VariableScrambler(mySecretKey)
  }

  @Provides
  @ApplicationScope
  internal fun providesAccountManager(): AccountManager {
    return AccountManager(androidAccountManager)
  }

  @Provides
  @ApplicationScope
  internal fun provideDatabase(application: Application): AppDatabase {
    return Room.databaseBuilder(application.applicationContext, AppDatabase::class.java,
        "motoq-db").build()
  }

}
